File: apps/api-go/internal/application/campaign_service.go
// Package application contains the services that orchestrate use cases. A
// service is a thin façade: it groups the use cases that belong to one area
// (campaigns) and exposes them as a single dependency for the HTTP handlers.
// Business rules stay in the use cases; the service only wires and forwards.
package application

import (
	"context"

	"github.com/minijuegosz1531/facebook_ads/apps/api-go/internal/domain/model"
	"github.com/minijuegosz1531/facebook_ads/apps/api-go/internal/domain/port"
	"github.com/minijuegosz1531/facebook_ads/apps/api-go/internal/domain/usecase"
)

// CampaignService groups every campaign-related operation behind one type.
type CampaignService struct {
	repo         port.CampaignRepository
	create       *usecase.CreateCampaignUseCase
	updateStatus *usecase.UpdateCampaignStatusUseCase
	insights     *usecase.GetInsightsUseCase
	publish      *usecase.SelectAssetsAndPublishUseCase
}

// NewCampaignService builds the service and all its use cases from the ports.
// The AdPlatform is passed per-call-site because in production it is bound to a
// specific client's ad account (see infrastructure.Container).
func NewCampaignService(adPlatform port.AdPlatform, repo port.CampaignRepository, jobStore port.JobStore) *CampaignService {
	create := usecase.NewCreateCampaignUseCase(adPlatform, repo)
	return &CampaignService{
		repo:         repo,
		create:       create,
		updateStatus: usecase.NewUpdateCampaignStatusUseCase(adPlatform, repo),
		insights:     usecase.NewGetInsightsUseCase(adPlatform, repo),
		publish:      usecase.NewSelectAssetsAndPublishUseCase(jobStore, create),
	}
}

func (s *CampaignService) Create(ctx context.Context, cmd usecase.CreateCampaignCommand) (model.Campaign, error) {
	return s.create.Execute(ctx, cmd)
}

func (s *CampaignService) PublishFromJob(ctx context.Context, cmd usecase.SelectAndPublishCommand) (model.Campaign, error) {
	return s.publish.Execute(ctx, cmd)
}

func (s *CampaignService) UpdateStatus(ctx context.Context, id string, status model.CampaignStatus) (model.Campaign, error) {
	return s.updateStatus.Execute(ctx, id, status)
}

func (s *CampaignService) Get(ctx context.Context, id string) (model.Campaign, error) {
	return s.repo.Get(ctx, id)
}

func (s *CampaignService) ListByClient(ctx context.Context, clientID string) ([]model.Campaign, error) {
	return s.repo.ListByClient(ctx, clientID)
}

func (s *CampaignService) Insights(ctx context.Context, id, since, until string) (map[string]any, error) {
	return s.insights.Execute(ctx, id, since, until)
}

func (s *CampaignService) Delete(ctx context.Context, id string) error {
	return s.repo.Delete(ctx, id)
}

File: apps/api/application/campaign_service.py
"""Application service orchestrating the campaign use cases."""
from __future__ import annotations

from domain.models.campaign import Campaign
from domain.ports.inbound.campaign_service import ICampaignService
from domain.ports.outbound.ad_platform import IAdPlatform
from domain.ports.outbound.campaign_repository import ICampaignRepository
from domain.ports.outbound.job_store import IJobStore
from domain.use_cases.create_campaign import CreateCampaignCommand, CreateCampaignUseCase
from domain.use_cases.get_insights import GetInsightsUseCase
from domain.use_cases.select_assets_and_publish import (
    SelectAndPublishCommand,
    SelectAssetsAndPublishUseCase,
)
from domain.use_cases.update_campaign_status import UpdateCampaignStatusUseCase


class CampaignService(ICampaignService):
    def __init__(
        self,
        ad_platform: IAdPlatform,
        campaign_repo: ICampaignRepository,
        job_store: IJobStore | None = None,
    ):
        self._repo = campaign_repo
        self._create = CreateCampaignUseCase(ad_platform, campaign_repo)
        self._update_status = UpdateCampaignStatusUseCase(ad_platform, campaign_repo)
        self._insights = GetInsightsUseCase(ad_platform, campaign_repo)
        self._publish = (
            SelectAssetsAndPublishUseCase(job_store, self._create)
            if job_store is not None
            else None
        )

    async def create_campaign(self, cmd: CreateCampaignCommand) -> Campaign:
        return await self._create.execute(cmd)

    async def publish_from_job(self, cmd: SelectAndPublishCommand) -> Campaign:
        if self._publish is None:
            raise RuntimeError("CampaignService was built without a job store")
        return await self._publish.execute(cmd)

    async def update_status(self, campaign_id: str, status: str) -> Campaign:
        return await self._update_status.execute(campaign_id, status)

    async def get(self, campaign_id: str) -> Campaign | None:
        return await self._repo.get(campaign_id)

    async def list_by_client(self, client_id: str) -> list[Campaign]:
        return await self._repo.list_by_client(client_id)

    async def get_insights(self, campaign_id: str, since: str, until: str) -> dict:
        return await self._insights.execute(campaign_id, since, until)

    async def delete(self, campaign_id: str) -> None:
        await self._repo.delete(campaign_id)
